go/merge/three_way.go

```diff
@@ -0,0 +1,259 @@
+			aChange = <-aChangeChan
+		}
+		if bChange.V == nil {
+			bChange = <-bChangeChan
+		}
+
+		// Both channels are producing zero values, so we're done.
+		if aChange.V == nil && bChange.V == nil {
+			break
+		}
+
+		// Since diff generates changes in key-order, and we never skip over a change without processing it, we can simply compare the keys at which aChange and bChange occurred to determine if either is safe to apply to the merge result without further processing. This is because if, e.g. aChange.V.Less(bChange.V), we know that the diff of b will never generate a change at that key. If it was going to, it would have done so on an earlier iteration of this loop and been processed at that time.
+		// It's also obviously OK to apply a change if only one diff is generating any changes, e.g. aChange.V is non-nil and bChange.V is nil.
+		if aChange.V != nil && (bChange.V == nil || aChange.V.Less(bChange.V)) {
+			merged = apply(merged, aChange, a.Get(aChange.V))
+			aChange = types.ValueChanged{}
+			continue
+		} else if bChange.V != nil && (aChange.V == nil || bChange.V.Less(aChange.V)) {
+			merged = apply(merged, bChange, b.Get(bChange.V))
+			bChange = types.ValueChanged{}
+			continue
+		}
+		d.PanicIfTrue(!aChange.V.Equals(bChange.V), "Diffs have skewed!") // Sanity check.
+
+		// If the two diffs generate different kinds of changes at the same key, conflict.
+		if aChange.ChangeType != bChange.ChangeType {
+			return parent, newMergeConflict("Conflict:\n%s\nvs\n%s\n", changeString(aChange), changeString(bChange))
+		}
+
+		aValue, bValue := a.Get(aChange.V), b.Get(bChange.V)
+		if aChange.ChangeType == types.DiffChangeRemoved || aValue.Equals(bValue) {
+			// If both diffs generated a remove, or if the new value is the same in both, merge is fine.
+			merged = apply(merged, aChange, aValue)
+		} else {
+			// There's one case that might still be OK even if aValue and bValue differ: different, but mergeable, compund values of the same type being added/modified at the same key, e.g. a Map being added to both a and b. If either is a primitive, or Values of different Kinds were added, though, we're in conflict.
+			if unmergeable(aValue, bValue) {
+				return parent, newMergeConflict("Conflict:\n%s = %s\nvs\n%s = %s", changeString(aChange), types.EncodedValue(aValue), changeString(bChange), types.EncodedValue(bValue))
+			}
+			// TODO: Add concurrency.
+			mergedValue, err := threeWayMerge(aValue, bValue, parent.Get(aChange.V), vwr)
+			if err != nil {
+				return parent, err
+			}
+			merged = merged.Set(aChange.V, mergedValue)
+		}
+		aChange, bChange = types.ValueChanged{}, types.ValueChanged{}
+	}
+	return merged, nil
+}
+
+func apply(target types.Map, change types.ValueChanged, newVal types.Value) types.Map {
+	switch change.ChangeType {
+	case types.DiffChangeAdded:
+		fallthrough
+	case types.DiffChangeModified:
+		return target.Set(change.V, newVal)
+	case types.DiffChangeRemoved:
+		return target.Remove(change.V)
```

Comment:
We can figure this out later - but every call to Set and Remove will re-chunk the map, and it'll get slow.